### DISCUSSION TITLE
Unlisted headers (\section*{}) equivalent

### DISCUSSION THREAD

**Discussion by jurf** [Witiko/markdown#511] [Q&A] [answered]
Hi,

is there a way to add headers that should be ignored by the ToC in markdown? All I found was #401 but not this. Unfortunately my current project needs them extensively, so I currently have to resort to raw `\section*{}` tags.

**Comment by Witiko** [comment, marked as answer]
Hi. After you have enabled [the option `headerAttributes`][1], you should be able to write either `{-}` or `{.unnumbered}` after headings:

``` tex
\documentclass{article}
\usepackage[headerAttributes]{markdown}
\begin{document}
\begin{markdown}
# Here is a numbered section
## Here is an unnumbered subsection {-}
### This subsection will be unnumbered as well
## Here is a numbered subsection
\end{markdown}
\end{document}
```

![image](https://github.com/user-attachments/assets/52d81e93-2994-43cf-a29d-93059f380dad)

 [1]: https://witiko.github.io/markdown/#header-attributes

**Reply by jurf** [reply]
Hi, thanks a lot for the quick reply! It was greatly appreciated, even though I did not get to replying at the time.

I think I figured out why I dismissed `unnumbered` - it does not work at all for me. I guess the `markdown` version in Overleaf (or TeX Live) is not new enough. :/ https://www.overleaf.com/read/kpbdzzttpjqh#8f7af1

**Reply by Witiko** [reply]
Sadly, Overleaf is ca 6 months out-of-date and still uses version 3.4.2 of the Markdown package from March while the latest released version is 3.7.1 from September.

If all you want are better defaults that support unnumbered sections out-of-box, you can include an up-to-date file `markdownthemewitiko_markdown_defaults.sty` in your Overleaf project, see also [my example Overleaf project](https://www.overleaf.com/read/dtgrncnfnhwq#1cd842).

However, new TeX interfaces were added to the Markdown package since version 3.4.2, most notably the support for incremental renderer (prototype) definitions, added in version 3.5.0 from April.

<details>

[![image](https://github.com/user-attachments/assets/139be900-ff63-4a1e-ab47-83e75778306f)][2]

</details>

The updated file `markdownthemewitiko_markdown_defaults.sty` uses the new interfaces for the new definitions of unnumbered sections. Therefore, they happen to work with version 3.4.2 only by a lucky accident. Furthermore, there have been many other new developments and bugfixes by me and @lostenderman since version 3.4.2, see also [the changelog][3].

Therefore, if you are serious about using the Markdown package on Overleaf, you may want to consider [installing the current version of the Markdown package][1] into your Overleaf project, see also [my example Overleaf project](https://www.overleaf.com/read/twfsstwwwgjf#3b0bc6). Alternatively, you may want to use an environment with an up-to-date TeX installation other than Overleaf, such as your laptop and the `texlive/texlive` Docker image.

 [1]: https://witiko.github.io/markdown/#portable-installation
 [2]: https://mirrors.ctan.org/macros/generic/markdown/markdown.pdf#paragraph.2.2.5.45
 [3]: https://github.com/Witiko/markdown/blob/main/CHANGES.md

**Reply by jurf** [reply]
Ah, I understand, thanks. Sorry, even after years with LaTeX, it can be overwhelming.

However, even after adding the latest version and successfully removing the numbers, the sections still show up in the ToC.

I updated the initial example: https://www.overleaf.com/read/kpbdzzttpjqh#8f7af1

**Reply by jurf** [reply]
Btw, I solved this for now by limiting the depth of the ToC which so far works in my current use-case.

**Reply by Witiko** [reply]
> However, even after adding the latest version and successfully removing the numbers, the sections still show up in the ToC.

This is as expected. The LaTeX command `\section*` is a low-level command that only formats a text to look like a heading without creating an entry in the table of contents, updating the running heads, creating a target for a hyperlink, or doing any of the other tasks that are necessary for a proper heading. By contrast, the default renderer for Markdown creates a regular heading, just without the numbering. In my experience, this is what is usually wanted in a document. However, I recognize that you asked specifically for a version that excludes the section from the table of contents.

If you'd like to translate `## Section name` to `\section*{Section Name}` instead, you can type the following in the preamble:

``` tex
\markdownSetup {
  renderers = {
    headingTwo = {\section*{#1}},
  },
}
```

If you'd like to do this conditionally just for `## Section name {-}` and not `## Section name`, then you need a more complex setup that would react to attribute values. Look for `\l__markdown_header_unnumbered_bool` in the file `markdownthemewitiko_markdown_defaults.sty` to see how it's done in the default definitions.

> Btw, I solved this for now by limiting the depth of the ToC which so far works in my current use-case.

I am glad that you found a solution that works for you.

**Reply by jurf** [reply]
> limiting the depth of the ToC which so far works in my current use-case

Turns out that wasn't enough so I tried creating the new renderer. I somehow managed to pound your instructions with a hammer until something useable came out of it. :) If anyone needs this in the future, this renderer seems to work. There is no guarantee, I really just added hashes at what looked like the most probable space until it started working.

```tex
\ExplSyntaxOn
\markdownSetup
  {
    rendererPrototypes = {
      headerAttributeContextBegin += {
        \markdownSetup
          {
            rendererPrototypes = {
              attributeClassName = {
                \bool_if:nT
                  {
                    \str_if_eq_p:nn
                      { ##1 }
                      { unnumbered } &&
                    ! \l__markdown_header_unnumbered_bool
                  }
                  {
                    \group_begin:
                    \bool_set_true:N
                      \l__markdown_header_unnumbered_bool
                    \markdownSetup
                      {
                        renderers = {
                          headingTwo = {\section*{####1}},
                          headingThree = {\subsection*{####1}},
                          headingFour = {\subsubsection*{####1}},
                        },
                        rendererPrototypes = {
                          sectionBegin = {
                            \group_begin:
                          },
                          sectionEnd = {
                            \group_end:
                          },
                        },
                      }
                  }
              },
            },
          }
      },
    },
  }
\ExplSyntaxOff
```

**Reply by jurf** [reply]
Hi, @Witiko, I am very sorry to bother you but I ran into a problem with this approach a few hours before handing over my thesis. Do you have any idea why this approach would break labels? It keeps complaining about multiple definitions, no matter what I do.

If it isn't obvious, no problem, I'll use the backup option and refer to the parent section, it's not too big of a deal.

**Reply by Witiko** [reply]
I am sorry to hear that!

> Do you have any idea why this approach would break labels?

Labels to the unnumbered sections themselves or to other material within these sections, such as tables and figures? If it's the sections, then you obviously wouldn't be able to refer to the headings created with `\section*`, `\subsection*`, and `\subsubsection*`, since they are unnumbered, i.e. there is no identifier to reference. If it's the other material, then what do the logs say? Can you isolate a section of the thesis text that reproduces the issue? Can you share your thesis?

**Reply by Witiko** [reply]
On a more philosophical note: Why is it only in the final hours before a submission that all these issues pop up? 😞

**Reply by jurf** [reply]
Hey, thanks a lot for the reply! I use the implementation above. It seems as though the identifiers do not get removed from the identifier list, and are applied until it is cleared. E.g., this:

```md
## Section {#sec:section}

### Unlisted subsection 1 {#sec:unlisted1 -}

### Unlisted subsection 2 {#sec:unlisted2 -}

### Other subsection with ID {#sec:regular}
```

Gives an output similar to this:
```
LaTeX: Label sec:unlisted1 is multiply defined
LaTeX: Label sec:unlisted1 is multiply defined
LaTeX: Label sec:unlisted2 is multiply defined
```

I've tried clearing it in several different places but so far I've only managed to make things worse. It's a nuisance as I can always just use sections and look a bit sloppy, but it's not an outright error.

The answer to the philosophical note is actually quite pragmatic. :) I cross-compile to HTML and use that mainly as it takes around 20 seconds to compile, not several minutes. The output seemed fine when I checked it, but I did not check the number. That is not to say I am not guity; if I had not been ignoring LaTeX compiler errors for the last two months I've might've noticed. 😬

**Reply by Witiko** [reply]
How interesting. I can reproduce the errors with the default definitions in the current version of the package:

``` tex
\documentclass{article}
\usepackage[headerAttributes]{markdown}
\begin{document}
\begin{markdown}

## Section {#sec:section}

### Unlisted subsection 1 {#sec:unlisted1 -}

### Unlisted subsection 2 {#sec:unlisted2 -}

### Other subsection with ID {#sec:regular}

\end{markdown}
\end{document}
```

``` log
LaTeX Warning: Label `sec:unlisted1' multiply defined.
LaTeX Warning: Label `sec:unlisted1' multiply defined.
LaTeX Warning: Label `sec:unlisted2' multiply defined.
```

Let's see what's going on.

**Reply by Witiko** [reply]
I see what's going on.

You should be able to hotfix the issue by adding the following code after `\usepackage{markdown}`:

``` tex
% Always clear the identifiers after a heading, globally.
\ExplSyntaxOn
\markdownSetup {
  rendererPrototypes = {
    headerAttributeContextEnd += {
      \seq_gclear:N
        \l__markdown_header_identifiers_seq
    }
  }
}
\ExplSyntaxOff
```

Please, let me know if this helped.

<details>

The issue is that we currently only clear the list of identifiers locally. However, this has a bad interaction with `{-}`, which starts a new group. The list of identifiers is cleared within this group but is restored after this group has ended.

For example, here are the elements produced by `### Unlisted subsection 1 {#sec:unlisted1 -}`:

- `sectionBegin`
- `headerAttributeContextBegin`: Redefines `attributeIdentifier` and `attributeClassName`.
- `attributeIdentifier{sec:unlisted1}`: Adds `sec:unlisted1` to `\l__markdown_header_identifiers_seq`.
- `attributeClassName{unnumbered}`: Starts a new group, enables `\l__markdown_header_unnumbered_bool`, and redefines `sectionBegin`/`sectionEnd` to start/end a group.
- `headingThree{Unlisted subsection 1}`: Renders `\subsection*{Unlisted subsection 1}`.
- `headerAttributeContextEnd`: Renders `\label{sec:unlisted}` and clears `\l__markdown_header_identifiers_seq`.
- `interblockSeparator`
- `sectionEnd`: Ends the current group, disables `\l__markdown_header_unnumbered_bool` (this, we want), and restores `sec:unlisted1` to `\l__markdown_header_identifiers_seq` (this, we don't).

The solution seems to be using a global variable `\g__markdown_header_identifiers_seq` and global assignments instead of the local variable `\l__markdown_header_identifiers_seq` and local assignments. Functionally, there should be no difference and we will become less vulnerable to TeX groups that are opened between `attributeIdentifier` and `headerAttributeContextEnd`.

After you have submitted your thesis and recovered from the crunch, perhaps you'd like to open a PR with this fixup for the next release of the Markdown package? :wink:  The variable is `\l__markdown_header_unnumbered_bool` named `\l_@@_header_identifiers_seq` in the file `markdown.dtx`, where the change should be made.

</details>

**Reply by jurf** [reply]
I owe you a beer.

**Reply by jurf** [reply]
> After you have submitted your thesis and recovered from the crunch, perhaps you'd like to open a PR with this fixup for the next release of the Markdown package? 😉 The variable is \l__markdown_header_unnumbered_bool named \l_@@_header_identifiers_seq in the file markdown.dtx, where the change should be made.

I would prefer the beer but I guess this is fair. 😅

**Reply by Witiko** [reply]
Glad I could help!

> I would prefer the beer but I guess this is fair. 😅

Contributing is its own reward—or so they say—so I still think you owe me a beer, no matter who ends up making the change. :wink:  Cheers!
